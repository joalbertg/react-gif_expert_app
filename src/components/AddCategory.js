import React, {useState} from 'react';
import PropTypes from 'prop-types';

const AddCategory = ({setCategories}) => {
  const [value, setValue] = useState('');
  const handleAdd = event => {
    setValue(event.target.value);
  }

  const handleSubmit = event => {
    event.preventDefault();

    if(value.trim().length > 0) {
      setCategories(categories => [...categories, value]);
      setValue('');
    }
  }

  return(
    <form onSubmit={handleSubmit}>
      <input
        type="text"
        value={value}
        onChange={handleAdd}
        placeholder='Add here...'
      />
    </form>
  );
}

AddCategory.propTypes = {
  setCategories: PropTypes.func.isRequired
}

export default AddCategory;

