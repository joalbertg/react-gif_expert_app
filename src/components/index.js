export { default as AddCategory } from './AddCategory';
export { default as ListCategories } from './ListCategories';
export { default as GifGrid } from './GifGrid';

