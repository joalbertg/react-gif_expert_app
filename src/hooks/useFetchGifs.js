import {useState} from 'react';

export const useFetchGifs = () => {
  const [state, setState] = useState({
    data: [],
    loading: true
  });

  setTimeout(() => {
    setState({
      ...state,
      loading: false
    })
  }, 3000);

  return state; // { data: [], loading: true }
}

