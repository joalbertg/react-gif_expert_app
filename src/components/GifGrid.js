import React from 'react';
import PropTypes from 'prop-types';

import { useFetchGifs } from '../hooks';
import GifGridItem from './GifGridItem';

const GifGrid = ({category}) => {
  const {data: gifs, loading} = useFetchGifs(category);

  return(
    <div className="card-grid">
      <h3 className="animate__animated animate__fadeIn">{category}</h3>
      { loading && <p className="loading animate__animated animate__flash">Loading</p> }
      {
        gifs.map(gif => (
          <GifGridItem
            key={gif.id}
            {...gif}
          />)
        )
      }
    </div>
  );
}

GifGrid.propTypes = {
  category: PropTypes.string.isRequired
}

export default GifGrid;

