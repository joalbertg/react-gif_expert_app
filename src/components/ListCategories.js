import React from 'react';

import GifGrid from './GifGrid';

const ListCategories = ({categories}) => {
  return(
    <>
      { categories ?
          categories.map(category => (
            <GifGrid
              key={category}
              category={category}
            />
          ))
        :
        <h3>No categories to display.</h3>
      }
    </>
  );
}

export default ListCategories;

