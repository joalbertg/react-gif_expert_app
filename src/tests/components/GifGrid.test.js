import React from 'react';
import { shallow } from 'enzyme';

import { GifGrid } from '../../components';
import { useFetchGifs } from '../../hooks/useFetchGifs';
// con esto hacemos mock del customHook
jest.mock('../../hooks/useFetchGifs');

describe('Tests GifGrid component', () => {
  const category = 'Dragon Ball';

  test('Should display correctly', () => {
    useFetchGifs.mockReturnValue({
      data: [],
      loading: true
    });

    const wrapper = shallow(<GifGrid category={category} />);
    expect(wrapper).toMatchSnapshot();
  });

  test('Should show items when loading gifs with useFetchGifs', () => {
    const gifs = [{
      id: 'ABC123',
      url: 'https://localhost:3000/something.gif',
      title: 'something'
    },
    {
      id: '123ABC',
      url: 'https://localhost:3000/something_v2.gif',
      title: 'something_v2'
    }];

    useFetchGifs.mockReturnValue({
      data: gifs,
      loading: false
    });

    const wrapper = shallow(<GifGrid category={category} />);

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('p').exists()).toBe(false);
    expect(wrapper.find('GifGridItem').length).toBe(gifs.length);
  });
});
