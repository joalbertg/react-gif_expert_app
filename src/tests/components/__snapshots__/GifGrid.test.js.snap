// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests GifGrid component Should display correctly 1`] = `
<div
  className="card-grid"
>
  <h3
    className="animate__animated animate__fadeIn"
  >
    Dragon Ball
  </h3>
  <p
    className="loading animate__animated animate__flash"
  >
    Loading
  </p>
</div>
`;

exports[`Tests GifGrid component Should show items when loading gifs with useFetchGifs 1`] = `
<div
  className="card-grid"
>
  <h3
    className="animate__animated animate__fadeIn"
  >
    Dragon Ball
  </h3>
  <GifGridItem
    id="ABC123"
    key="ABC123"
    title="something"
    url="https://localhost:3000/something.gif"
  />
  <GifGridItem
    id="123ABC"
    key="123ABC"
    title="something_v2"
    url="https://localhost:3000/something_v2.gif"
  />
</div>
`;
